channel.go

```diff
@@ -1370,8 +1410,24 @@ PublishWithDeferredConfirm behaves identically to Publish but additionally retur
 DeferredConfirmation, allowing the caller to wait on the publisher confirmation
 for this message. If the channel has not been put into confirm mode,
 the DeferredConfirmation will be nil.
+
+Deprecated: Use PublishWithDeferredConfirmWithContext instead.
 */
 func (ch *Channel) PublishWithDeferredConfirm(exchange, key string, mandatory, immediate bool, msg Publishing) (*DeferredConfirmation, error) {
+	return ch.PublishWithDeferredConfirmWithContext(context.Background(), exchange, key, mandatory, immediate, msg)
+}
+
+/*
+PublishWithDeferredConfirmWithContext behaves identically to Publish but additionally returns a
+DeferredConfirmation, allowing the caller to wait on the publisher confirmation
+for this message. If the channel has not been put into confirm mode,
+the DeferredConfirmation will be nil.
+*/
+func (ch *Channel) PublishWithDeferredConfirmWithContext(ctx context.Context, exchange, key string, mandatory, immediate bool, msg Publishing) (*DeferredConfirmation, error) {
+	if ctx == nil {
+		return nil, errors.New("amqp091-go: nil Context")
+	}
```

Comment:
I think this check is not needed.
If somebody passes a nil context, `context.WithCancel` will panic and the user will get the same error with a stacktrace instead.

It's more graceful to return an error then to panic but you won't get a stacktrace which makes it harder to debug.
The ctx parameter is also not really different then others parameter. We also don't check if other parameters are nil or have a value.
Therefore I think the check can be removed.

Reply:
I hesitate on this part and based myself on what `net/http` does in a similar context as an example and they generate an error on WithContext methods with a `nil` context: https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/net/http/request.go;l=856-868 to prevent later panic. But I agree that both solution have their advantages and disadvantages.